### PR TITLE
ddl: label constraint support single quotes

### LINKED ISSUE #20668: ddl: label constraint doesn't support single quotes
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
mysql> alter table t1 alter partition p1 alter placement policy constraints="['+zone=sh']" role=voter replicas=3;
```

### 2. What did you expect to see? (Required)

It should not report the error, as it's very common to replace double quotes with single quotes.

### 3. What did you see instead (Required)

```
ERROR 8234 (HY000): Invalid placement policy 'alter placement policy constraints='[''+zone=sh'']' role=voter replicas=3': invalid character '\'' looking for beginning of value
```

### 4. What is your TiDB version? (Required)

master

## Score

- 300

## Mentor

- @djshow832

### PR DESCRIPTION
**Pull request by pengdaqian2020**
### What problem does this PR solve?

Issue Number: close #20668

Problem Summary:

### What is changed and how it works?

What's Changed:

Replace single quotes with double quotes

How it Works:

Replace single quotes with double quotes

### Check List 

Tests

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
No
    - Consumes more MEM
No
- Breaking backward compatibility
No

### Release note

- Replace single quotes with double quotes

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -5778,6 +5778,7 @@ func buildPlacementSpecReplicasAndConstraint(replicas uint64, cnstr string) ([]*
 
 		constraints := []string{}
 
+		cnstr = strings.Replace(cnstr, "'", "\"", -1)
```
Comment (djshow832) [inline ddl/ddl_api.go:5781]: It's not right.
e.g. `['+zone=sh"]`

Comment (pengdaqian2020) [reply ddl/ddl_api.go]: fixed

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -5777,8 +5777,35 @@ func buildPlacementSpecReplicasAndConstraint(replicas uint64, cnstr string) ([]*
...
+					leftIndex = 1
+				case len(byteArr[i]) - 1:
+					rightIndex = len(byteArr[i]) - 2
+				default:
+					leftIndex, rightIndex = 0, len(byteArr[i])-1
+				}
+				if len(byteArr) == 1 {
+					leftIndex, rightIndex = 1, len(byteArr[i])-2
+				}
+				if byteArr[i][leftIndex] == '\'' {
```
Comment (djshow832) [inline ddl/ddl_api.go:5798]: White spaces may also appear here.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -5777,8 +5777,35 @@ func buildPlacementSpecReplicasAndConstraint(replicas uint64, cnstr string) ([]*
 		}
 
 		constraints := []string{}
+		byteArr := bytes.Split([]byte(cnstr), []byte(","))
```
Comment (djshow832) [inline ddl/ddl_api.go:5780]: Commas may be in the quotes.

**Review by djshow832 — COMMENTED** [review]
Both `[]` and `{}` formats need to support quotes.

File: ddl/ddl_api.go (hunk 2 above)
Comment (pengdaqian2020) [reply ddl/ddl_api.go]: Done

File: ddl/ddl_api.go (hunk 3 above)
Comment (pengdaqian2020) [reply ddl/ddl_api.go:5780]: if commas may be in the quotes, it will parse error, right ?

Comment (pengdaqian2020) [reply ddl/ddl_api.go:5780]: the code will deal with this situation
```
if len(byteArr[i]) == 0 {
	break
}
```

**Review by xhebox — COMMENTED** [review]
The code is to replace single quotes with double quotes to adapt json standard.
I would recommend to try other yaml/toml libs. I did not find a suitable library than `json.Unmarshal` when I am coding.
To replace quotes in a robust way, you will need to fully cover a partial json syntax, in a systematic way like DFA. If your code knows how to correctly replace quotes, I believe it is already ready to replace `json.Unmarshal`.

**Comment by xhebox** [comment]
#20668 has been closed

**Pull request closed without merging**